Cleanup and organization

richardkmichael opened the pull request:
Rename and move several functions for clarity.

Moving OpenCL platform init inside the CasOFFinder object in [6c15532](https://github.com/snugel/cas-offinder/commit/6c15532b9b80b68934be6e0060a0c41fcacc93b3) is debatable.  At this time, there is only one class instance so the instance constructor could do the init.  However, if multiple instances are intended to be possible then presumably platform init must not by done by the instance constructor and each instance will be responsible for devices only.

I can remove this commit, if so.

richardkmichael committed:
Add comment explaining SNUCL

richardkmichael committed:
Remove unused `maxdevnum`

richardkmichael committed:
Move `releaseLociinfo()` to honour main call-sequence for readability

richardkmichael committed:
README typos

richardkmichael committed:
Rename `init_platforms()` to `initOpenCLPlatforms()` for clarity

richardkmichael committed:
Add comments related to parsing input

richardkmichael committed:
Rename `initOpenCL()` to `initOpenCLDevices()` for clarity

richardkmichael committed:
Clean up indentation

richardkmichael committed:
README formatting and typos

richardkmichael committed:
Mention optional OpenMP

richardkmichael committed:
Improve CMake installation instructions

richardkmichael committed:
Provide cmake help instructions

richardkmichael committed:
Wrap paragraphs for editing convenience

richardkmichael committed:
Consistent indentation

richardkmichael committed:
Indentation whitespace

richardkmichael committed:
Move `add_executable()` to provide a target for later use

richardkmichael committed:
Bump cmake version to use recent modules and targets

richardkmichael committed:
Use target and import for OpenCL, Apple special handling no longer required

richardkmichael committed:
Use imported OpenMP provider with target

richardkmichael committed:
Remove include directory, handled in `config.h`

richardkmichael committed:
Consistent spacing

richardkmichael committed:
Group package declarations

richardkmichael committed:
Adjust module path near usage

richardkmichael committed:
Perform GNU compiler check early

richardkmichael committed:
Write config.h last

richardkmichael committed:
Remove used module include

richardkmichael committed:
Add homepage to CMake project and use in `print_usage()`

richardkmichael commented:
Note, this is on top of #36

pjb7687 commented on CMakeLists.txt:
```diff
@@ -1,41 +1,11 @@
-cmake_minimum_required (VERSION 3.1)
-project (Cas-OFFinder)
+cmake_minimum_required (VERSION 3.19.1)
```
Is the newest version needed? I prefer keeping the oldest version here to make it compatible with old systems.

richardkmichael replied:
I am sure the newest version is not needed, and I agree with the idea of preserving compatibility for old systems.  On the other hand, I do not know CMake well, so I am uncertain how far back it is possible to go (and have the improved modules work as required here).

TL;DR I will roll back to 3.14 and see what happens.  FWIW, I think it could use latest and it would not be too much of a problem.  More info below.

I have just read all the release notes between 3.1 and 3.19.

In 3.10, cmake itself dropped support for C++98 compilers.  (So if you are on a system which requires a C++98 compiler, then you cannot compile cmake >= 3.10.  Unlikely situation, I believe.)

In 3.14, cmake dropped support for WindowsXP and Vista.

In 3.15, cmake requires XCode 5 for the XCode generator.
  (This is not relevant here; I suppose anyone doing current macOS/iOS development is already > XCode5 based on Apple itself, but also, multiple cmake versions could be installed if < XCode 5 support is required for a different project.)

In 3.19, cmake dropped compatibility with build systems generates by cmake < 2.8.12.  (Not relevant here, and again an additional, old version of cmake could be installed by people with other, old cmake projects.  I do not think we need to worry about this.)

I think cmake support of old systems is quite good, and updating to latest cmake will not leave behind very many (maybe zero) users.

So, it probably comes down to "install convenience" as .msi / .exe, DMG, vendor .deb/.rpm.

Windows users will presumably download binaries from KitWare, so installing the latest version is easy for them.

macOS users can install the latest cmake easily via HomeBrew, and are probably accustomed to regular updates in any case.  (I myself find this frustrating though!)  HomeBrew cmake does not depend on any non-system packages, so updating cmake should not be a problem (i.e., no updates of other packages would be required).

macOS users can also download binaries from KitWare, support for macOS > 10.10.

Ubuntu Linux users can use KitWare's [APT repo](https://apt.kitware.com/), which supports Ubuntu 16.04, 18.04, 20.04.  This handles the case where Ubuntu itself does not update cmake.

For common platforms, this leaves Fedora Linux.  I am not sure how far back Fedora will issue updates, but I guess it is not very far since Fedora moves quickly.

pjb7687 replied:
Then let's keep it 3.1 ATM. This is the oldest release supports `set (CMAKE_CXX_STANDARD 11)`.

richardkmichael replied:
Ok, I will check if the changes I made work with 3.1.

I do think it would be good to update, at least partly.  There are lots of fixes in cmake on each release and it is a fairly complex build system.

pjb7687 commented on cas-offinder.cpp:
```diff
@@ -181,7 +190,7 @@ void Cas_OFFinder::setChrData() {
 				((m_chrdatasize / m_devnum) + ((m_chrdatasize%m_devnum == 0) ? 0 : 1))
 			)
 		); // No more than maximum allocation per device
-		// cerr << "Dicesize: " << m_dicesizes[dev_index] << endl;
```
Would you please keep them? These are useful for debugging purposes.

pjb7687 commented on cas-offinder.cpp:
```diff
@@ -440,19 +430,31 @@ void Cas_OFFinder::compareAll(const char* outfilename) {
 void Cas_OFFinder::print_usage() {
 	unsigned int i, j;
-	cout << "Cas-OFFinder 3.0.0b (" << __DATE__ << ")" << endl <<
+	cout << "Cas-OFFinder v" << CAS_OFFINDER_VERSION << " (Compiled on: " << __DATE__ << ")" << endl <<
```
I consider removing the prefix 'v' from the version number from 3.0 release

richardkmichael replied:
Sorry I missed this!

I removed the `v` on the `develop-build-changes` changes of #36, and rebased here too.

pjb7687 commented on main.cpp:
```diff
@@ -10,17 +10,16 @@
 #include <cstdlib>
 #include <cstdarg>
 
-//#define DEBUG
```
Also keep it, please!

richardkmichael replied:
Sure, I will deal with the DEBUG related code shortly.  Thanks!

pjb7687 commented:
> Rename and move several functions for clarity.
> 
> Moving OpenCL platform init inside the CasOFFinder object in [6c15532](https://github.com/snugel/cas-offinder/commit/6c15532b9b80b68934be6e0060a0c41fcacc93b3) is debatable. At this time, there is only one class instance so the instance constructor could do the init. However, if multiple instances are intended to be possible then presumably platform init must not by done by the instance constructor and each instance will be responsible for devices only.
> 
> I can remove this commit, if so.

I think it makes sense to call init outside the object. IIRC they are static methods for that purpose.

pjb7687 commented:
And seems like this PR is duplicate #36 which one should I merge?

richardkmichael commented:
> And seems like this PR is duplicate #36 which one should I merge?

As I mentioned in a comment above, this PR is on top of #36.  I left this one WIP: so you would not merge it (and therefore accidentally bring in #36).

I will reply to your comments, inline above.

Thank you for the fast review!

richardkmichael commented:
> > Rename and move several functions for clarity.
> > Moving OpenCL platform init inside the CasOFFinder object in [6c15532](https://github.com/snugel/cas-offinder/commit/6c15532b9b80b68934be6e0060a0c41fcacc93b3) is debatable. At this time, there is only one class instance so the instance constructor could do the init. However, if multiple instances are intended to be possible then presumably platform init must not by done by the instance constructor and each instance will be responsible for devices only.
> > I can remove this commit, if so.
> 
> I think it makes sense to call init outside the object. IIRC they are static methods for that purpose.

Ok, I dropped that commit.

richardkmichael commented:
Sorry for the PR confusion.

I realize now that GitHub does not support the `WIP:` prefix to prevent merging and it also doesn't allow a PR to be based on another PR in an obvious way.

GitLab supports both of these.  This makes it easier to send all work for review, and allow the relationships to be clear.

richardkmichael committed:
Bump cmake version to use recent modules and targets

richardkmichael committed:
Add version to CMake project and use in `print_usage()`

richardkmichael commented:
@pjb7687

Ok, I've addressed all your comments!

I think there is more elegant way to handle debugging, but I need to learn more about cmake.  I do that in the future in a separate PR.

I believe if you merge this PR, it will also automatically close #36 (since this branch is on top of the branch in #36).

richardkmichael commented:
@pjb7687 

Ok, I believe I've addressed all your comments, thanks!

I need to learn more about cmake for a more elegant solution to debugging, and will investigate in the future.

I believe if you merge this PR, then #36 will also close automatically, since this branch is on top of the #36 branch.
